src/ir/module-splitting.cpp

```diff
@@ -0,0 +1,405 @@
+  auto secondary = std::make_unique<Module>();
+  secondary->features = primary.features;
+  secondary->hasFeaturesSection = primary.hasFeaturesSection;
+  return secondary;
+}
+
+void shareImportableItems(Module& primary,
+                          Module& secondary,
+                          Name importNamespace) {
+  // Map internal names to (one of) their corresponding export names.
+  std::unordered_map<Name, Name> exports;
+  for (auto& ex : primary.exports) {
+    if (ex->kind != ExternalKind::Function) {
+      exports[ex->value] = ex->name;
+    }
+  }
+
+  auto makeImportExport = [&](Importable& primaryItem,
+                              Importable& secondaryItem,
+                              Name exportName,
+                              ExternalKind kind) {
+    secondaryItem.name = primaryItem.name;
+    secondaryItem.hasExplicitName = primaryItem.hasExplicitName;
+    secondaryItem.module = importNamespace;
+    auto exportIt = exports.find(primaryItem.name);
+    if (exportIt != exports.end()) {
+      secondaryItem.base = exportIt->second;
+    } else {
+      exportName = Names::getValidExportName(primary, exportName);
+      primary.addExport(new Export{exportName, primaryItem.name, kind});
+      secondaryItem.base = exportName;
+    }
+  };
+
+  // TODO: Be more selective by only sharing global items that are actually used
+  // in the secondary module, just like we do for functions.
+
+  auto shareMemory = [&](Memory* memory) {
+    secondary.memory.exists = true;
+    secondary.memory.initial = memory->initial;
+    secondary.memory.max = memory->max;
+    secondary.memory.shared = memory->shared;
+    secondary.memory.indexType = memory->indexType;
+    makeImportExport(*memory, secondary.memory, "memory", ExternalKind::Memory);
+  };
+  ModuleUtils::iterImportedMemories(primary, shareMemory);
+  ModuleUtils::iterDefinedMemories(primary, shareMemory);
+
+  auto shareTable = [&](Table* table) {
+    secondary.table.exists = true;
+    secondary.table.initial = table->initial;
+    secondary.table.max = table->max;
+    makeImportExport(*table, secondary.table, "table", ExternalKind::Table);
+  };
+  ModuleUtils::iterImportedTables(primary, shareTable);
+  ModuleUtils::iterDefinedTables(primary, shareTable);
+
+  auto shareGlobal = [&](Global* global) {
+    assert(primary.features.hasMutableGlobals() &&
+           "TODO: add wrapper functions for disallowed mutable globals");
+    auto secondaryGlobal = std::make_unique<Global>();
+    secondaryGlobal->type = global->type;
+    secondaryGlobal->mutable_ = global->mutable_;
+    secondaryGlobal->init =
+      global->init == nullptr
+        ? nullptr
+        : ExpressionManipulator::copy(global->init, secondary);
+    makeImportExport(*global, *secondaryGlobal, "global", ExternalKind::Global);
+    secondary.addGlobal(std::move(secondaryGlobal));
+  };
+  ModuleUtils::iterImportedGlobals(primary, shareGlobal);
+  ModuleUtils::iterDefinedGlobals(primary, shareGlobal);
```

Comment:
is there a reason to have a helper function and these two calls, and not just
```cpp
for (auto& global : module->globals)
```
to iterate over them all? Seems like we do the same for imported and defined ones anyhow.